File: src/robot_debug/opengl_gui/widget.cpp
#include <iostream>

#include "widget.h"

#include "widget_label.h"
#include "widget_frame.h"
#include "widget_text_frame.h"
#include "widget_model_frame.h"

Widget::Widget()
{

}

Widget::Widget(GLVisualisation &visualisation_, Variables &variables_, Json::Value &params_)
{
  init(visualisation_, variables_, params_);
}

void Widget::init(GLVisualisation &visualisation_, Variables &variables_, Json::Value &params_)
{
  visualisation = &visualisation_;
  variables     = &variables_;
  params        = params_;

  m_type = params["type"].asString();

  x = params["position"][0].asFloat();
  y = params["position"][1].asFloat();
  z = params["position"][2].asFloat();

  r = params["color"][0].asFloat();
  g = params["color"][1].asFloat();
  b = params["color"][2].asFloat();

  auto variable_name  = params["variable"]["name"].asString();
  auto variable_value = params["variable"]["value"];
  variables->add(variable_name, variable_value);

  visible_on(); 

  std::cout << "creating widget " << params["type"].asString() << "\n";

  for (unsigned int i = 0; i < params["widgets"].size(); i++)
  {
    std::string type = params["widgets"][i]["type"].asString();

    if (type == "label")
      widgets.push_back(new WidgetLabel(visualisation_, variables_, params_["widgets"][i]));
    else
    if (type == "frame")
      widgets.push_back(new WidgetFrame(visualisation_, variables_, params_["widgets"][i]));
    else
    if (type == "text frame")
      widgets.push_back(new WidgetTextFrame(visualisation_, variables_, params_["widgets"][i]));
    else
    if (type == "model frame")
      widgets.push_back(new WidgetModelFrame(visualisation_, variables_, params_["widgets"][i]));
    else
      widgets.push_back(new Widget(visualisation_, variables_, params_["widgets"][i]));
  }
}

Widget::Widget(Widget& other)
{
  copy(other);
}

Widget::Widget(const Widget& other)
{
  copy(other);
}

Widget::~Widget()
{
  for (unsigned int i = 0; i < widgets.size(); i++)
    delete widgets[i];
}

Widget& Widget::operator= (Widget& other)
{
  copy(other);
  return *this;
}

Widget& Widget::operator= (const Widget& other)
{
  copy(other);
  return *this;
}


void Widget::copy(Widget& other)
{
  visualisation = other.visualisation;
  variables     = other.variables;
  params        = other.params;
  m_visible     = other.m_visible;

  m_type        = other.m_type;

  x = other.x;
  y = other.y;
  z = other.z;

  r = other.r;
  g = other.g;
  b = other.b;
}

void Widget::copy(const Widget& other)
{
  visualisation = other.visualisation;
  variables     = other.variables;
  params        = other.params;
  m_visible     = other.m_visible;

  m_type        = other.m_type;

  x = other.x;
  y = other.y;
  z = other.z;

  r = other.r;
  g = other.g;
  b = other.b;
}



void Widget::render()
{
  if (m_visible)
  {
    visualisation->push();

    visualisation->translate(0, 0, 0);
    visualisation->set_color(1, 1, 1);

    visualisation->print(0, 0, 0, "default widget");

    for (unsigned int i = 0; i < widgets.size(); i++)
      widgets[i]->render();


    visualisation->pop();

  }
}

void Widget::visible_on()
{
  m_visible = true;
}

void Widget::visible_off()
{
  m_visible = false;
}

File: src/tools/robot_debug_py/opengl_gui/widget.py



class Widget:
    def __init__(self, visualisation, variables, textures, params):
        self.visualisation = visualisation
        self.variables     = variables
        self.textures      = textures
        self.params        = params


        self.m_type = str(self.params["type"])

        k = 0.5
        self.x = float(self.params["position"][0])*k
        self.y = float(self.params["position"][1])*k
        self.z = float(self.params["position"][2])*k

        self.r = float(self.params["color"][0])
        self.g = float(self.params["color"][1])
        self.b = float(self.params["color"][2])

        self.font_size = float(self.params["font_size"])


        if "variable" in self.params:
            self.variable_name  = str(self.params["variable"]["name"])
            self.variable_value = self.params["variable"]["value"]
            self.variables.add(self.variable_name, self.variable_value)
        else:
            self.variable_name  = None
            self.variable_value = None

        self.visible_on()

        print("creating widget ", str(self.params["type"]))

        self.child_widgets = []

        from opengl_gui.widget_frame import WidgetFrame
        from opengl_gui.widget_bar_graph import WidgetBarGraph
        from opengl_gui.widget_text_frame import WidgetTextFrame
        from opengl_gui.widget_distance_sensor import WidgetDistanceSensor
        from opengl_gui.widget_model_frame import WidgetModelFrame
        from opengl_gui.widget_camera_frame import WidgetCameraFrame

        if "widgets" in self.params:
            for i in range(len(self.params["widgets"])):
                widget_type = self.params["widgets"][i]["type"]

                if widget_type == "frame":
                    widget = WidgetFrame(visualisation, variables, textures, params["widgets"][i])
                elif widget_type == "text_frame":
                    widget = WidgetTextFrame(visualisation, variables, textures, params["widgets"][i])    
                elif widget_type == "bar_graph":
                    widget = WidgetBarGraph(visualisation, variables, textures, params["widgets"][i])
                elif widget_type == "distance_sensors":
                    widget = WidgetDistanceSensor(visualisation, variables, textures, params["widgets"][i])
                elif widget_type == "model_frame":
                    widget = WidgetModelFrame(visualisation, variables, textures, params["widgets"][i])
                elif widget_type == "camera_frame":
                    widget = WidgetCameraFrame(visualisation, variables, textures, params["widgets"][i])
                else:
                    widget = Widget(visualisation, variables, textures, params["widgets"][i])

                self.child_widgets.append(widget)



    def visible_on(self):
        self.visible = True

    def visible_off(self):
        self.visible = False

    def render(self):
        if self.visible:
            self.visualisation.push()

            for i in range(len(self.child_widgets)):
                self.child_widgets[i].render()

            self.visualisation.pop()